src/Collections/Arrays/BinarySearch.dfy

```diff
@@ -0,0 +1,40 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../Wrappers.dfy"
+include "../../Relations.dfy"
+
+module BinarySearch {
+  import opened Wrappers
+  import opened Relations
+
+  method BinarySearch<T>(a: array<T>, key: T, less: (T, T) -> bool) returns (r: Option<nat>)
+    requires SortedBy(a[..], (x, y) => less(x, y) || x == y)
```

Comment:
This will have the same issue I commented on for `Trichotomous`, making it impossible to sort by projections. Would it be possible to make the proof go through if we required `TotalOrdering` instead of `StrictTotalOrdering`?

Reply:
I attempted that at first, but I was discouraged not only by verification not immediately working, but also the fact that it makes the specification messier as well: instead of 

```dafny
ensures r.Some? ==> r.value < a.Length && a[r.value] == key
```

You would need

```dafny
ensures r.Some? ==> r.value < a.Length && lessThanOrEqual(a[r.value], key) && lessThanOrEqual(key, a[r.value])
```

Which is just plain much harder to read and understand.

I think for that reason, it makes more sense to improve this at the same time that we add comparison functions that return `CmpResult` instead as per #53.